### ISSUE TITLE
Option to never close a window on child death even with no background/disowned processes

### ISSUE THREAD

**Issue by lilyball** [kovidgoyal/kitty#4036]
**Is your feature request related to a problem? Please describe.**
If I run a program in a new Kitty window that prints output and immediately exits, Kitty will show the window, run the program, and immediately close the window. This happens regardless of `close_on_child_death` as that only keeps the window open if a disowned/background process is still kicking around.

This is frustrating because it means I can't quickly open a new window with the output of some command without wrapping it in a shell that waits for input before terminating. This is doubly frustrating if my intent is to run a command with precisely the initial terminal environment and don't want e.g. `/bin/sh` to change anything.

It also means if I accidentally kill a process (e.g. hit Ctrl-D in my shell when I didn't mean to) I immediately lose all scrollback, and I also lose the notion of "current directory" (e.g. I can't spawn a new shell).

**Describe the solution you'd like**
I'd like an option to have windows stay open after their process died. They should be visibly marked as having no process, but should not close until I explicitly close them. I should be able to set this in kitty.conf, and perhaps override it with a flag when invoking `kitty` directly (e.g. if I'm opening a window to an editor, as ctrl+shift+f2 does, I may want it to close when the editor dies).

Precedent here is Apple Terminal, which does precisely this. When the last process in a window ends, the text `[Process completed]` is printed to the terminal and the window and tab titles are updated to remove the process name. Also, very importantly, the notion of "current directory" actually persists, so if I open a new tab, it keeps the same directory.

**Describe alternatives you've considered**
There's room for subtlety here. Kitty could do something like "if the user hits Ctrl-D and that causes the process to terminate, go ahead and close the window", or "if the process was using alternate screen mode, it was likely interactive so go ahead and close the window" (this latter one makes things like `kitty vim` work nicely out of the box). I don't know what exact logic would go here, and maybe this is something that would want to also be configurable (e.g. let me say "be intelligent about it" versus "never close").

**Additional context**
I don't know how this would handle splits. Apple Terminal doesn't have splits so it doesn't have to deal with that, it just has tabs and windows. It's certainly possible this could be handled as "splits will close immediately, but a tab or OS window won't close without using the keyboard shortcut for it", though that's a bit odd.

Ultimately, it's very surprising to me for a terminal window to just vanish. It's also very annoying that I can't run e.g. `kitty env` to see the initial shell environment (this is something I actually do in Apple Terminal more often than you'd think, mostly when mucking about with my shell initialization in order to verify what it sees on startup).

There's also a feature that could be unlocked with this functionality, which is a way to say "please respawn my shell", or rather "please rerun the command that was used to spawn this window in the first place". That itself has questions like "if we preserve the current directory, what if the command had a relative path as an argument?", but mostly it would just be a really quick way of saying "oops I didn't mean to exit my shell, please spawn a new shell in the current directory, ideally without losing my scrollback". Any customized shell environment would be lost, but it would still be quite useful.

**Comment by lilyball** [comment]
Side note: Does Kitty even support the OSC 7 escape for telling the terminal what the current directory is? It doesn't show that info anywhere I can see, and Fish doesn't include Kitty in the list of terminals it sends OSC 7 to.

**Comment by kovidgoyal** [comment]
Staying open after child death
kitty --hold

# Showing env vars
https://sw.kovidgoyal.net/kitty/actions/#show-kitty-env-vars

# Using the current directory to open new windows
https://sw.kovidgoyal.net/kitty/launch/#cmdoption-launch-cwd

# Display the current directory in tab/window title without needing hacks like OSC 7
https://github.com/kovidgoyal/kitty/issues/3948

**Comment by lilyball** [comment]
Can you please stop summarily closing my issues without actually understanding my problem?

`--hold` is a flag to the `kitty` binary. It is not a config option, and it only applies to the very first window opened using `kitty --hold`. This means it won't apply to any new windows/tabs created within the running Kitty instance. That means it won't apply to 99% of my Kitty windows.

I didn't know about this flag originally, so it does handle the specific case of `kitty --hold env`, but it does not handle the more general case of "the command I was running exited but I didn't want that to happen and now I've lost the whole window".

> `show_kitty_env_vars`

This looks useful, thanks.

> https://sw.kovidgoyal.net/kitty/launch/#cmdoption-launch-cwd

This does not solve my problem. I've already rebound keys to the `_use_cwd` variants. The problem I described was once the window has closed I no longer have the cwd to create new shells from.

> Display the current directory in tab/window title without needing hacks like OSC 7

What makes OSC 7 a hack? If anything, "show the cwd by setting the window title" is the hack. Several issues:

* If I'm running a command, I lose the display of my current directory because it's replaced by the command
* Overriding shell RC files automatically is not great. I realize your docs have a "if you don't want this, here's how you set it up manually" but still.
* That solution does not scale to any shells not explicitly supported by Kitty. OSC 7 is something any shell can start printing without requiring the terminal emulator to know about that shell.
* OSC 7 takes a file:/// URI, which allows terminals to do things like show a proxy icon in the titlebar. Setting the window title to the path does not have any semantic meaning.
* It overrides any existing `fish_title` I might have already configured. I can set `no-title` but then I'm back to not having the path in my window title.
* It duplicates the fish truncated prompt into the window title, which may be either too verbose or not verbose enough. In apps like Terminal.app, besides the proxy icon it also means the window title contains just the current dir name but cmd-clicking will give me a dropdown with the full path (which is standard OS behavior for windows with proxy icons).
* The use of a file:// URI qualified with hostname allows the proxy icon to even understand if this is a local path or a remote path.

Basically, OSC 7 is the right way to report the cwd, and the only thing hacky about it is the lack of proper feature detection (though is there even any downside to sending it to terminal emulators that don't handle it, assuming they handle OSC codes at all?).

This is rather orthogonal to this particular issue though.

**Comment by kovidgoyal** [comment]
On Wed, Sep 15, 2021 at 07:38:15PM -0700, Lily Ballard wrote:
> Can you please stop summarily closing my issues without actually understanding my problem?

I manage the triage for issues in my project, how I like. You are free
to manage triage for issues in your projects how you like. An issue
being closed does not mean you cannot continue the discussion.

> 
> `--hold` is a flag to the `kitty` binary. It is not a config option, and it only applies to the very first window opened using `kitty --hold`. This means it won't apply to any new windows/tabs created within the running Kitty instance. That means it won't apply to 99% of my Kitty windows.
> 
> I didn't know about this flag originally, so it does handle the specific case of `kitty --hold env`, but it does not handle the more general case of "the command I was running exited but I didn't want that to happen and now I've lost the whole window".

And for that case we have: sh -c "command; read". 

> 
> > `show_kitty_env_vars`
> 
> This looks useful, thanks.
> 
> > https://sw.kovidgoyal.net/kitty/launch/#cmdoption-launch-cwd
> 
> This does not solve my problem. I've already rebound keys to the `_use_cwd` variants. The problem I described was once the window has closed I no longer have the cwd to create new shells from.

Once the window has closed there is nothing there to launch from. I am
guessing you mean once the child has died.

> 
> > Display the current directory in tab/window title without needing hacks like OSC 7
> 
> What makes OSC 7 a hack? If anything, "show the cwd by setting the window title" is the hack. Several issues:

Already discussed to death. Find th proposal for it in terminal-wg

> 
> * If I'm running a command, I lose the display of my current directory because it's replaced by the command

That would be the same for osc 7 and has nothing to do with the
mechanism for getting that information to the terminal.

> * Overriding shell RC files automatically is not great. I realize your docs have a "if you don't want this, here's how you set it up manually" but still.

Again has nothing to do with OSC 7.

> * That solution does not scale to any shells not explicitly supported by Kitty. OSC 7 is something any shell can start printing without requiring the terminal emulator to know about that shell.

Huh? This mechanism works with all shells, non-shells and *any process*
at all, without it needing to do anything. The terminal uses the actual
existing CWD of the process as the CWD of the window, without the
process needing to do anything.

> * OSC 7 takes a file:/// URI, which allows terminals to do things like show a proxy icon in the titlebar. Setting the window title to the path does not have any semantic meaning.

I dont what a proxy icon is, probably yet another poorly thought out
Apple hack.

> * It overrides any existing `fish_title` I might have already configured. I can set `no-title` but then I'm back to not having the path in my window title.

So set your own fish_title that displays the CWD, there is zero need for
OSC 7 to do that.

> * It duplicates the fish truncated prompt into the window title, which may be either too verbose or not verbose enough. In apps like Terminal.app, besides the proxy icon it also means the window title contains just the current dir name but cmd-clicking will give me a dropdown with the full path (which is standard OS behavior for windows with proxy icons).

Yup, Apple-ism.

> * The use of a file:// URI qualified with hostname allows the proxy icon to even understand if this is a local path or a remote path.

And how would it do anything with remote URLs?

> 
> Basically, OSC 7 is the right way to report the cwd, and the only thing hacky about it is the lack of proper feature detection (though is there even any downside to sending it to terminal emulators that don't handle it, assuming they handle OSC codes at all?).

No, it isnt. 
https://gitlab.freedesktop.org/terminal-wg/specifications/-/issues/20

> 
> This is rather orthogonal to this particular issue though.

Indeed, it is.

**Comment by kovidgoyal** [comment]
Oh and before I forget, if you dont like running shells to hold commands, you can do

kitty +hold command

instead of

sh -c "command; read"

**Comment by lilyball** [comment]
> Once the window has closed there is nothing there to launch from. I am guessing you mean once the child has died.

Exactly my point! This is what my whole issue is about!

Once the window has closed, there is no longer a container that was associated with a cwd, and therefore no way to open a new window with the same cwd.

If the window does not go away immediately, it can remember what the cwd was I can make a new window/tab using that. This is what e.g. Terminal.app does. If I exit my process, the window doesn't close immediately, and New Tab will create a new tab using the same cwd.

Having said that, experimentally right now using `kitty --hold`, kitty does not in fact use the cwd from the process that just exited. If I run `kitty --hold`, change directories, hit Ctrl-D to exit the shell, and then my hotkey for `new_window_with_cwd`, the new window has kitty's cwd rather than the cwd of the shell that just exited. I consider this a bug.

> Already discussed to death. Find th proposal for it in terminal-wg

Found it. There's no clear consensus. One person (jerch) is very aggressively against this feature. Everyone else seems to be taking it more seriously.

> Huh? This mechanism works with all shells, non-shells and *any process* at all, without it needing to do anything. The terminal uses the actual existing CWD of the process as the CWD of the window, without the process needing to do anything.

Of which process? If there's just one, sure. What if there's a tree of processes? Which one gets to determine the cwd to show in the terminal? It's typically the shell that you want, but what if there are multiple shells (or the terminal emulator doesn't know which process is the shell). Picking whatever seems to be the most recent interactive process doesn't work either (`bash -c 'cd foo && less bar.txt'` shouldn't cause `less` to define the cwd).

> I dont what a proxy icon is, probably yet another poorly thought out Apple hack.

You literally have no idea what it is and yet you're declaring it to be a poorly-thought-out hack. I guess I shouldn't be surprised given how extremely antagonistic you have been to every single piece of feedback I have tried to give you.

---

Given how hostile you are to making any changes to behavior whatsoever, I give up. You win. I'll stop filing feedback (and just stop trying to use Kitty altogether). Congratulations, this is the second-worst OSS experience I've ever had. Give yourself a medal.

**Comment by kovidgoyal** [comment]
On Wed, Sep 15, 2021 at 08:48:58PM -0700, Lily Ballard wrote:
> > Once the window has closed there is nothing there to launch from. I am guessing you mean once the child has died.
> 
> Exactly my point! This is what my whole issue is about!
> 
> Once the window has closed, there is no longer a container that was associated with a cwd, and therefore no way to open a new window with the same cwd.

This would be true using OSC 7 as well.

> 
> If the window does not go away immediately, it can remember what the cwd was I can make a new window/tab using that. This is what e.g. Terminal.app does. If I exit my process, the window doesn't close immediately, and New Tab will create a new tab using the same cwd.
> 
> Having said that, experimentally right now using `kitty --hold`, kitty does not in fact use the cwd from the process that just exited. If I run `kitty --hold`, change directories, hit Ctrl-D to exit the shell, and then my hotkey for `new_window_with_cwd`, the new window has kitty's cwd rather than the cwd of the shell that just exited. I consider this a bug.

I dont. 

> 
> > Already discussed to death. Find th proposal for it in terminal-wg
> 
> Found it. There's no clear consensus. One person (jerch) is very aggressively against this feature. Everyone else seems to be taking it more seriously.

That discussion lists all the problems with OSC 7. 

> 
> > Huh? This mechanism works with all shells, non-shells and *any process* at all, without it needing to do anything. The terminal uses the actual existing CWD of the process as the CWD of the window, without the process needing to do anything.
> 
> Of which process? If there's just one, sure. What if there's a tree of processes? Which one gets to determine the cwd to show in the terminal? It's typically the shell that you want, but what if there are multiple shells (or the terminal emulator doesn't know which process is the shell). Picking whatever seems to be the most recent interactive process doesn't work either (`bash -c 'cd foo && less bar.txt'` shouldn't cause `less` to define the cwd).

The foreground process. So if you open vim and cd somewhere else in vim,
it will still work, unlike with the hack that is OSC 7.

> 
> > I dont what a proxy icon is, probably yet another poorly thought out Apple hack.
> 
> You literally have no idea what it is and yet you're declaring it to be a poorly-thought-out hack. I guess I shouldn't be surprised given how extremely antagonistic you have been to every single piece of feedback I have tried to give you.

I said probably, and given I am intimately familiar with all the pain
Apple and its notions have caused me over the years, that is a very high probability.

> 
> ---
> 
> Given how hostile you are to making any changes to behavior whatsoever, I give up. You win. I'll stop filing feedback (and just stop trying to use Kitty altogether). Congratulations, this is the second-worst OSS experience I've ever had. Give yourself a medal.

You reap what you sow. You want decent OSS experiences, start by not
assuming you know better than everyone else and that everyone has to do
what you tell them to, just because you and your corporate overlord
have said so.
